output/driver_unix.go

```diff
@@ -183,16 +193,96 @@ func (out *output) openAndSetupMixer() error {
 		return out.alsaError("snd_mixer_selem_get_playback_volume_range", err)
 	}
 
-	// set initial volume and verify it actually works
-	mixerVolume := out.volume*(float32(out.mixerMaxVolume-out.mixerMinVolume)) + float32(out.mixerMinVolume)
-	if err := C.snd_mixer_selem_set_playback_volume_all(out.mixerElemHandle, C.long(mixerVolume)); err != 0 {
-		return out.alsaError("snd_mixer_selem_set_playback_volume_all", err)
+	if out.externalVolume {
+
+		// get current volume from the mixer, and set the spotify volume accordingly
+		var volume C.long
+		C.snd_mixer_selem_get_playback_volume(out.mixerElemHandle, C.SND_MIXER_SCHN_MONO, &volume)
+		out.volume = float32(volume-out.mixerMinVolume) / float32(out.mixerMaxVolume-out.mixerMinVolume)
+
+		out.externalVolumeUpdate <- out.volume
+	} else {
+
+		// set initial volume and verify it actually works
+		mixerVolume := out.volume*(float32(out.mixerMaxVolume-out.mixerMinVolume)) + float32(out.mixerMinVolume)
+		if err := C.snd_mixer_selem_set_playback_volume_all(out.mixerElemHandle, C.long(mixerVolume)); err != 0 {
+			return out.alsaError("snd_mixer_selem_set_playback_volume_all", err)
+		}
 	}
 
+	// set callback and initialize private
+	var cb C.snd_mixer_elem_callback_t = (C.snd_mixer_elem_callback_t)(C.alsaMixerCallback)
+	C.snd_mixer_elem_set_callback(out.mixerElemHandle, cb)
+	C.snd_mixer_elem_set_callback_private(out.mixerElemHandle, unsafe.Pointer(&out.volume))
+
+	go out.waitForMixerEvents()
+
 	out.mixerEnabled = true
 	return nil
 }
 
+func (out *output) waitForMixerEvents() {
+	for !out.closed {
+		var res = C.snd_mixer_wait(out.mixerHandle, -1)
+		if out.closed {
+			// if we reach here, the playing context has probably changed
+			break
+		}
+		if res >= 0 {
+			res = C.snd_mixer_handle_events(out.mixerHandle)
+			if res <= 0 {
+				// todo add warning
+				continue
+			}
+
+			var priv = float32(*(*C.float)(C.snd_mixer_elem_get_callback_private(out.mixerElemHandle)))
+			if priv < 0 {
+				// volume update came from spotify, so no need to tell spotify about it
+				// reset the private, but discard the event
+				C.snd_mixer_elem_set_callback_private(out.mixerElemHandle, unsafe.Pointer(&out.volume))
+
+				continue
+			}
+			if priv == out.volume {
+				log.Debugf("volume repeated. skipping event (volume=%f)\n", priv)
```

Comment:
I'd replace this with something more verbose like `skipping ALSA mixer event, volume already updated: %.2f`